baseapp/abci_utils.go

```diff
@@ -330,3 +312,74 @@ func NoOpVerifyVoteExtensionHandler() sdk.VerifyVoteExtensionHandler {
 		return &abci.ResponseVerifyVoteExtension{Status: abci.ResponseVerifyVoteExtension_ACCEPT}, nil
 	}
 }
+
+// TxSelector defines a helper type that assists in selecting transactions during
+// mempool transaction selection in PrepareProposal. It keeps track of the total
+// number of bytes and total gas of the selected transactions. It also keeps
+// track of the selected transactions themselves.
+type TxSelector struct {
+	maxTxBytes   uint64
+	maxBlockGas  uint64
+	totalTxBytes uint64
+	totalTxGas   uint64
+	selectedTxs  [][]byte
+}
+
+func NewTxSelector(maxTxBytes, maxBlockGas uint64) *TxSelector {
+	return &TxSelector{
+		maxTxBytes:  maxTxBytes,
+		maxBlockGas: maxBlockGas,
+	}
+}
+
+// SelectedTxs returns a copy of the selected transactions.
+func (ts *TxSelector) SelectedTxs() [][]byte {
+	txs := make([][]byte, len(ts.selectedTxs))
+	copy(txs, ts.selectedTxs)
+	return txs
+}
+
+// Clear clears the TxSelector, nulling out all fields.
+func (ts *TxSelector) Clear() {
+	ts.totalTxBytes = 0
+	ts.totalTxGas = 0
+	ts.selectedTxs = nil
+}
+
+// SelectTxForProposal selects a transaction for inclusion in a proposal. It will
+// only select the provided transaction if there is enough capacity in the block.
+// It will return <true> if the caller should halt the transaction selection loop
+// (typically over a mempool) or <false> otherwise.
+func (ts *TxSelector) SelectTxForProposal(memTx sdk.Tx, txBz []byte) bool {
+	txSize := uint64(len(txBz))
+
+	var txGasLimit uint64
+	if memTx != nil {
+		if gasTx, ok := memTx.(GasTx); ok {
+			txGasLimit = gasTx.GetGas()
+		}
+	}
+
+	// only add the transaction to the proposal if we have enough capacity
+	if (txSize + ts.totalTxBytes) <= ts.maxTxBytes {
+		// If there is a max block gas limit, add the tx only if the limit has
+		// not been met.
+		if ts.maxBlockGas > 0 {
+			if (txGasLimit + ts.totalTxGas) <= ts.maxBlockGas {
+				ts.totalTxGas += txGasLimit
+				ts.totalTxBytes += txSize
+				ts.selectedTxs = append(ts.selectedTxs, txBz)
+			}
+		} else {
+			ts.totalTxBytes += txSize
+			ts.selectedTxs = append(ts.selectedTxs, txBz)
+		}
```

Comment:
We could invert these conditions to ease reading even better and make the structure even more obvious:
```go
if ts.maxBlockGas <= 0 { // No limit was defined here
    ts.totalTxBytes += txSize
    ts.selectedTxs = append(ts.selectedTxs, txBz)
} else if (txGasLimit + ts.totalTxGas) <= ts.maxBlockGas { // Within max block gas limit
    ts.totalTxGas += txGasLimit
    ts.totalTxBytes += txSize
   ts.selectedTxs = append(ts.selectedTxs, txBz)
}
```

Reply:
@alexanderbez did you dismiss this review or was there a change left?